### PR TITLE
added a metric monitoring Proto/Borsh encoding for peer connections

### PR DESCRIPTION
**Pull request by pompon0**
We will disable borsh encoding once node all versions >= PEER_MIN_ALLOWED_PROTOCOL_VERSION support protobuf decoding. Nevertheless, this metric will allow us to observe whether protobuf is actually used (as borsh is still enabled as a fallback).

### CODE REVIEW

**Comment by mina86** [comment]
This approach sounds very suspicious to me. Surely, we cannot disable Borsh encoding so long as PEER_MIN_ALLOWED_PROTOCOL_VERSION uses Borsh.  And really we should be detecting whether other peer supports proto by the protocol version they send in the `Handshake` which should always be encoded using Borsh.  But perhaps I’m missing context where this has been discussed.

**Comment by pompon0** [comment]
@mina86 design: https://docs.google.com/document/d/1gCWmt9O-h_-5JDXIqbKxAaSS3Q9pryB1f9DDY1mMav4/edit

> which should always be encoded using Borsh

FYI the "Borsh" part is exactly the thing that we want to replace.

**Review by jakmeier — COMMENTED** [review]
Hm, I had a look through the code and this change on its own looks ok to me.

But I think I am not the right code owner to have a look at this. (Never looked really looked at Prometheus stats nor the networking layer...)

@mina86  since you already joined the conversation, could you please give the code owner approval, once you approve the PR? (I think it's only about stats, the general migration strategy should be a different discussion.)

**Comment by mina86** [comment]
> FYI the "Borsh" part is exactly the thing that we want to replace.

Yes, I understand that.  But I don’t like the idea of sending proto to nodes which may not understand them.  PEER_MIN_ALLOWED_PROTOCOL_VERSION defines minimal protocol we support and so we should assume that node we’re talking to might use that version of the network protocol.  For the time being, Handshake should be sent as Borsh and then we can check protocol_version to see whether other messages should use other encodings.

Eventually, as we continue to increase STABLE_PROTOCOL_VERSION until PEER_MIN_ALLOWED_PROTOCOL_VERSION reaches version when proto was introduced we can start Handshakes as protobuffers.  If the receiving node is unable to parse that it’s a node we would be unable to talk to anyway.

There would be some ambiguity when receiving the first message from a node, but that can be trivially resolved by checking first byte of the serialised message: if it’s zero it’s a Borsh-serialised handshake. If it’s non-zero it’s protobuf message or a massage which is not accepted at this stage of the connection.

Furthermore, ‘We will disable the borsh encoding once we see in Prometheus that (almost) all connections in mainnet are using protobuf.’ explicitly breaks promise of PEER_MIN_ALLOWED_PROTOCOL_VERSION.

By the way, sending messages as proto breaks bunch of pytests: https://nayduck.near.org/#/run/2486

File: chain/network/src/peer/peer_actor.rs
Comment (pompon0) [reply chain/network/src/peer/peer_actor.rs:228]: the node should be able to gracefully handle garbage data from the peer, since it is an external source of data. We log parsing error as "debug", and definitely shouldn't log it as "error". What is the difference if we discard a message before or after decoding? IMO decoding a message that we intend to discard increases a risk of introducing a bug in which we start interpreting the message before discarding.

**Comment by pompon0** [comment]
> But I don’t like the idea of sending proto to nodes which may not understand them. PEER_MIN_ALLOWED_PROTOCOL_VERSION defines minimal protocol we support and so we should assume that node we’re talking to might use that version of the network protocol.

Sure, I agree with that. But I don't understand why you think that "sending a message that will be discarded by the receiver" cannot be a regular part of the protocol. We do not implement linear logic here, so discarding part of the input is a totally valid piece of logic.

> There would be some ambiguity when receiving the first message from a node, but that can be trivially resolved by checking first byte of the serialised message: if it’s zero it’s a Borsh-serialised handshake. If it’s non-zero it’s protobuf message or a massage which is not accepted at this stage of the connection.

This is exactly what we do already: the first byte of the message allows us to distinguish the encoding (see the doc). No need for extra migration steps. The code doing that might just be not as explicit as it could, but it is there.

> Furthermore, ‘We will disable the borsh encoding once we see in Prometheus that (almost) all connections in mainnet are using protobuf.’ explicitly breaks promise of PEER_MIN_ALLOWED_PROTOCOL_VERSION.

No, it doesn't if we bump PEER_MIN_ALLOWED_PROTOCOL_VERSION together with the deprecation. Let me update the PR description.

> By the way, sending messages as proto breaks bunch of pytests: https://nayduck.near.org/#/run/2486

ack, on it

File: chain/network/src/peer/peer_actor.rs
Comment (jakmeier) [reply chain/network/src/peer/peer_actor.rs:228]: Ok if we treat any de serialization errors as a pure debug info, then this seems okay.

**Comment by pompon0** [comment]
> So why do we need this metrics?

I've implemented prototobuf support with a fallback to borsh in case of problems. It is not enough to know that the newest version is present everywhere, we also need to know that it is actually using protobufs.

**Review by mina86 — APPROVED** [review]
The code here looks good but the commit message is still baffling to me.

Monitoring whether we’re actually using protobufs may be useful, but ‘we will disable the borsh encoding once we see in Prometheus that (almost) all connections in mainnet are using protobuf’ is still inappropriate approach. 

And ‘we will increase PEER_MIN_ALLOWED_PROTOCOL_VERSION appropriately when disabling borsh support’ is backwards. Min allowed version is always two less than current stable version. It’s that number which dictates when we can disable support for Borsh encoding not the other way around.  We may *also* look at the metric to make sure that everyone is indeed talking proto, but it cannot effect min protocol version increase. 

It’s likely that a week after 1.27.0 release all nodes will talk protobuf. You’re advocating that we should then increase PEER_MIN_ALLOWED_PROTOCOL_VERSION to 54 and have 1.28.0 with STABLE_PROTOCOL_VERSION == PEER_MIN_ALLOWED_PROTOCOL_VERSION == 54.

Then again, I guess I don’t care enough.

**Comment by pompon0** [comment]
Fair enough. Changed the description.

**Pull request merged**
